### PR TITLE
[query] add `pc_relate_ts` method (tall skinny implementation)

### PR DESCRIPTION
**Pull request by pwc2**
Adds the tall skinny (table of NDArrays) implementation of `pc_relate` and associated tests.

### CODE REVIEW

File: hail/python/hail/methods/relatedness/pc_relate_ts.py [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+    # Recover V from pc_scores with inv(S0)
+    V0 = (pc_scores * (1 / S0))._persist()
+    # First column in V needs all entries as 1/sqrt(n_samples), for intercept in beta
+    ones_normalized = hl.nd.full((V0.shape[0], 1), (1 / S[0]))
+    V = hl.nd.hstack((ones_normalized, V0))
+
+    nan = hl.float64(float("NaN"))
+
+    beta = g.select(ndarray=((1 / S) * V).T @ g.ndarray.T)
+    mu = beta.select(ndarray=((0.5 * V * S) @ beta.ndarray).T.map(lambda mu: hl.if_else(_bad_mu(mu, min_individual_maf), nan, mu)))
```
Comment (patrick-schultz) [inline hail/python/hail/methods/relatedness/pc_relate_ts.py:436]: It looks like from this point on, `beta` `S` and `V` aren't used again, only `mu`. Is that right?

You shouldn't need to actually materialize the `hstack`s. From some quick scratch work, it looks like you can equivalently use
```
g = g.annotate(af=g.ndarray.sum(1) / (2 * n))
mu = g.select(ndarray=g.af + 0.5 * g.ndarray @ V0 @ V0.T).map(lambda mu: hl.if_else(_bad_mu(mu, min_individual_maf), nan, mu)))
```
which feels more intuitive to me as well (mean center, compute PCs and project, then add the mean back). Happy to go through the math with you.

File: hail/python/hail/methods/relatedness/pc_relate_ts.py [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+    ----------
+    A : :class:`.Table`
+
+    Returns
+    -------
+    :class:`.NDArrayNumericExpression`
+        `A.T @ A`
+"""
+    A = A.select(ndarray=A.ndarray.T @ A.ndarray)
+    return A.aggregate(hl.agg.ndarray_sum(A.ndarray), _localize=False)
```
[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/python/hail/methods/relatedness/pc_relate_ts.py:62]: Even better, we might want to add an aggregator that uses the blas function `dsyrk` for computing gram matrices. It should be about a factor of 2 faster than the `dgemm` call.

**Review by patrick-schultz — CHANGES_REQUESTED** [review]
This is so cool, great work! Here are some initial thoughts.

File: hail/python/hail/methods/relatedness/pc_relate_ts.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (johnc1231) [reply hail/python/hail/methods/relatedness/pc_relate_ts.py:62]: Yeah, that is better. Should we expose a `gram` method or the like that calls that thing besides having it in an aggregator? I feel like it's common operation, but it's so much more readable to just write `A.T @ A`. Maybe it's a simplify rule we should add?

[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/python/hail/methods/relatedness/pc_relate_ts.py:62]: Yeah, I think we should expose it explicitly so experts can ensure the efficient implementation is used, while also making a best effort to optimize all code. In this case, I guess just looking for the pattern `ApplyAggOp(..., Seq(NDArrayTranspose(Ref(x)), Ref(x)), NDArrayMultiplySum)`.

**Comment by danking** [comment]
@pwc2 , I'm gonna close this because PR has trouble with old PRs. Can we revisit this when the PCA project is wrapped up?

**Comment by pwc2** [comment]
> @pwc2 , I'm gonna close this because PR has trouble with old PRs. Can we revisit this when the PCA project is wrapped up?

Yeah, sounds good

**Pull request closed without merging**